File: src/com/ianhanniballake/contractiontimer/appwidget/AppWidgetUpdateHandlerV11.java
package com.ianhanniballake.contractiontimer.appwidget;

import android.annotation.TargetApi;
import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;

import com.ianhanniballake.contractiontimer.R;

/**
 * Handles updating all App Widgets
 */
public class AppWidgetUpdateHandlerV11 extends AppWidgetUpdateHandlerBase
{
	/**
	 * Updates all instances of the Detail App Widgets
	 * 
	 * @param context
	 *            Context used to trigger updates
	 * @param appWidgetManager
	 *            AppWidgetManager instance
	 */
	@TargetApi(11)
	private static void updateDetailWidgets(final Context context, final AppWidgetManager appWidgetManager)
	{
		final int[] detailAppWidgetIds = appWidgetManager.getAppWidgetIds(new ComponentName(context,
				DetailAppWidgetProvider.class));
		final boolean detailWidgetsExist = detailAppWidgetIds.length > 0;
		if (detailWidgetsExist)
		{
			context.startService(new Intent(context, DetailAppWidgetService.class));
			appWidgetManager.notifyAppWidgetViewDataChanged(detailAppWidgetIds, R.id.list_view);
		}
	}

	@Override
	public void updateAllWidgets(final Context context)
	{
		super.updateAllWidgets(context);
		final AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
		AppWidgetUpdateHandlerV11.updateDetailWidgets(context, appWidgetManager);
	}
}

File: mobile/src/main/java/com/ianhanniballake/contractiontimer/appwidget/AppWidgetUpdateHandlerV11.kt
package com.ianhanniballake.contractiontimer.appwidget

import android.annotation.TargetApi
import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.os.Build
import com.ianhanniballake.contractiontimer.R
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

/**
 * Handles updating all App Widgets
 */
class AppWidgetUpdateHandlerV11 : AppWidgetUpdateHandlerBase() {
    /**
     * Updates all instances of the Detail App Widgets
     *
     * @param context          Context used to trigger updates
     * @param appWidgetManager AppWidgetManager instance
     */
    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    private suspend fun updateDetailWidgetsAsync(
            context: Context,
            appWidgetManager: AppWidgetManager
    ) = coroutineScope {
        async {
            val appWidgetIds = appWidgetManager.getAppWidgetIds(
                    ComponentName(context, DetailAppWidgetProvider::class.java))
            if (appWidgetIds.isNotEmpty()) {
                DetailAppWidgetProvider.updateDetailAppWidget(context, appWidgetIds)
                appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetIds, R.id.list_view)
            }
        }
    }

    override suspend fun collectWidgetUpdateAsync(
            context: Context,
            appWidgetManager: AppWidgetManager
    ) = listOf(
            *super.collectWidgetUpdateAsync(context, appWidgetManager).toTypedArray(),
            updateDetailWidgetsAsync(context, appWidgetManager))
}
